src/server/utils/watchLocalModules.js

```diff
@@ -17,54 +17,165 @@
+// * https://forums.docker.com/t/file-system-watch-does-not-work-with-mounted-volumes/12038
+// * runner on macOS chokidar would see the first change but not subsequent changes, presumably due
+//   to inode tracking and the way the build directories are destroyed and recreated
+// * using runner with --envVars.CHOKIDAR_USEPOLLING=true had no effect
+// for us the pattern of the files we want to check have a very consistent pattern and we don't need
+// to know the changes in the sometimes very many other files, so we can skip a lot of work from a
+// general solution
+
+async function changeHandler(changedPath) {
+  // I had apprehension of this manual manipulation
+  // but from what I can tell a file or directory with '/' in the name is not valid
+  // Linux: https://stackoverflow.com/q/9847288
+  // macOS: '/' from Finder is translated to ':'
+  // Windows: https://learn.microsoft.com/en-us/windows/win32/fileio/naming-a-file#naming-conventions
+  const [moduleName] = changedPath.split(path.sep);
+  const moduleMap = getModuleMap();
+  const moduleMapEntry = moduleMap.getIn(['modules', moduleName]);
+
+  if (!moduleMapEntry) {
+    console.warn(`module "${moduleName}" not in the module map, make sure to serve-module first`);
+    return;
+  }
+
+  console.log(`the Node.js bundle for ${moduleName} finished saving, attempting to load`);
+
+  let newModule;
+  try {
+    // FIXME: this leads to a race condition later
+    // (two modules changed at the same time, both editing different copies of the module map)
+    newModule = addHigherOrderComponent(await loadModule(
+      moduleName,
+      moduleMapEntry.toJS(),
+      onModuleLoad
+    ));
+  } catch (error) {
+    // loadModule already logs the error and then re-throws
+    // logging again just adds noise for the developer
+    return;
+  }
+
+  const newModules = getModules().set(moduleName, newModule);
+  resetModuleRegistry(newModules, moduleMap);
+  console.log(`finished reloading ${moduleName}`);
+}
 
 export default function watchLocalModules() {
   const staticsDirectoryPath = path.resolve(__dirname, '../../../static');
-  const moduleDirectory = path.resolve(staticsDirectoryPath, 'modules');
-  const moduleMapPath = path.resolve(staticsDirectoryPath, 'module-map.json');
-  const watcher = chokidar.watch(moduleDirectory, { awaitWriteFinish: true });
+  const moduleDirectory = path.join(staticsDirectoryPath, 'modules');
+
+  // this may be an over-optimization in that it may be more overhead than it saves
+  const stating = new Map();
+  function stat(filePath) {
+    if (!stating.has(filePath)) {
+      stating.set(
+        filePath,
+        fs.stat(filePath)
+          .then((fileStat) => {
+            stating.delete(filePath);
+            return fileStat;
+          })
+      );
+    }
 
-  watcher.on('change', async (changedPath) => {
-    try {
-      if (!changedPath.endsWith('.node.js')) return;
+    return stating.get(filePath);
+  }
```

Comment:
too much? (remove?)

Reply:
removed in [7398793](https://github.com/americanexpress/one-app/pull/1368/commits/73987933a31da700d91eb06fef6a359b35da3220)